### ISSUE TITLE
SQLNormalizedCache reactivity differs from LruNormalizedCache with watchers and prettifyDump

### ISSUE THREAD

**Issue by Nealsoni00** [apollographql/apollo-android#2693]
**Summary**

1. **Issue 1: Reactivity between two caches**

I'm experiencing issues with reactivity when using `LruNormalizedCache` when compared against `SQLNormalizedCache`. 

When using just `SQLNormalizedCache`, all watchers are called correctly when data is returned from the server after mutations and queries (just like the web and iOS clients). The interfaces all update correctly and the app launches using only the SQL cached data (indicating there is data in cache — see **Issue 2**).

However, when I use `LruNormalizedCache`, the watchers are not called when they should be in response to mutations (and are with SQLNormalizedCache). Any insight on how I can get the same cache reactivity behavior from the two would be super helpful!

2. **Issue 2: Direct cache access & prettifyDump**

When using just `SQLNormalizedCache`, I want to directly access a query and add data to that query (pagination of a list using infinite scroll). However, when I try to access the cache directly using `apolloClient.apolloStore.read(targetQuery).execute()` where `targetQuery` was executed at app launch and should be in cache, I get an `ApolloException: Failed to perform store operation`. 

When trying to dump the cache, it logs nothing and logging `apolloStore.normalizedCache().dump()` yields
```{class com.apollographql.apollo.cache.normalized.OptimisticNormalizedCache={}, class com.apollographql.apollo.cache.normalized.sql.SqlNormalizedCache={}}```

Even when I query using NETWORK_ONLY, then in the onResponse, I call the same exact query using CACHE_ONLY, I get a `null` response. This happens for both `SQLNormalizedCache` and `LruNormalizedCache`.

This is even more confusing considering the app launches correctly with no internet connection using only the `SQLNormalizedCache`. 

**Conclusion** 

Any insight on how to get the `apolloStore.read(targetQuery)` to return data and/or getting the `normalizedCache().dump()` to work would be greatly appreciated. Otherwise, insight on how to get the `LruNormalizedCache` to react the same way that the SQL cache does would also work. 

**Version**
build.gradle (:app)
```
apollo {
    generateKotlinModels = true
    customTypeMapping = [ 'JSON' : "org.json.JSONObject" ]
}

dependencies {
    def apolloVersion = "2.4.0"
    implementation "com.apollographql.apollo:apollo-runtime:$apolloVersion"
    implementation "com.apollographql.apollo:apollo-coroutines-support:$apolloVersion"
    implementation "com.apollographql.apollo:apollo-normalized-cache:$apolloVersion"
    implementation "com.apollographql.apollo:apollo-android-support:$apolloVersion"
    implementation "com.apollographql.apollo:apollo-normalized-cache-sqlite:$apolloVersion"
}
```

Apollo Cache Config
```
    private val resolver: CacheKeyResolver = object : CacheKeyResolver() {
        override fun fromFieldRecordSet(field: ResponseField, recordSet: Map<String, Any>): CacheKey {
            val id = recordSet["id"] as? String
            if (id != null) return CacheKey.from(id)
            return CacheKey.NO_KEY
        }

        override fun fromFieldArguments(field: ResponseField, variables: Operation.Variables): CacheKey {
            val id = field.resolveArgument("id", variables) as? String
            if (id != null) return CacheKey.from(id)
            return CacheKey.NO_KEY
        }
    }
```

```
private val cacheFactory = LruNormalizedCacheFactory(EvictionPolicy.builder().maxSizeBytes(10 * 1024 * 1024).build())
private val sqlCacheFactory = SqlNormalizedCacheFactory(context, "apollo.db")

apolloClient = ApolloClient.builder()
            .serverUrl(serverUrl)
            .logger(LocalLogger())
            .okHttpClient(okHttpClient)
//          .normalizedCache(cacheFactory, resolver) // Swap between two cache factories. This one is not reactive. 
            .normalizedCache(sqlCacheFactory, resolver)
            .addCustomTypeAdapter(CustomType.JSON, jsonCustomTypeAdapter)
            .defaultHttpCachePolicy(HttpCachePolicy.CACHE_FIRST)
            .build()
```

**Comment by martinbonnin** [comment]
Hi! Thanks for reaching out! Indeed, the normalized cache is an area that requires **a lot** of polishing/rework. It's pretty high in the todo list. Until then, things you can try:

* What is the size of the data you are downloading? Your `LruNormalizedCache` is configured for 10MB so there's a chance some items are evicted, which might explain why the cache is not reactive anymore.
* `SqlNormalizedCache` does not implement `dump` which explains why you aren't seeing anything. I created a follow up issue there: https://github.com/apollographql/apollo-android/issues/2696
* What you can try for debugging is to chain a `Lru` cache to a `SQL` cache:

```kotlin
LruNormalizedCacheFactory(EvictionPolicy.NO_EVICTION).chain(SqlNormalizedCacheFactory(this, "apollo.db"));
```

This way you'll be able to call dump and have all your cache (since there's no eviction)

* Other things to look for are cache misses due to missing ids even though it doesn't seem like it is the case here.

For the reactivity/`.read` not working issues, if you can upload a reproducer project somewhere, that'd help a ton. 
Hope this helps,
Martin

**Comment by Nealsoni00** [comment]
Thank you for your quick response, just wanted to update you before I close the issue. 
1. Chaining the caches does not work for the watchers or cache reads to work. I think it's the issue u described that `LruNormalizedCache` is not ready for production watchers. (They all work on SQL cache). The size of the data stored in cache is a maximum of 1MB in my cases.
2.  Thanks for adding #2696 will be useful when it's implemented. 
3. The read issues disappeared when I reinstalled the app from scratch on the 2 test devices and simulators which is super weird (It was happening on all devices I was testing on but then I created a new simulator and it worked).  Potentially it was a change to the db schema for SQLight?

Looking forward for the updates to the normalized cache!
